compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -3360,7 +3371,33 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                 }
 
                 match arg.layout.abi {
-                    Abi::Aggregate { .. } => {}
+                    Abi::Aggregate { .. } => {
+                        // Pass and return structures up to 2 pointers in size by value,
+                        // matching `ScalarPair`. LLVM will usually pass these in 2 registers
+                        // which is more efficient than by-ref.
+                        let ptr_size = Pointer.size(self);
+                        let max_by_val_size = ptr_size * 2;
+                        let size = arg.layout.size;
+
+                        if arg.layout.is_unsized() || size > max_by_val_size {
+                            arg.make_indirect();
+                        } else if size > ptr_size && self.has_all_float(&arg.layout) {
+                            // We don't want to aggregate floats as an aggregates of Integer
+                            // because this will hurt the generated assembly (#93490) but as an
+                            // optimization we want to pass homogeneous aggregate of floats
+                            // greater than pointer size as indirect.
+                            arg.make_indirect();
+                        } else {
+                            // We want to pass small aggregates as immediates, but using
+                            // a LLVM aggregate type for this leads to bad optimizations,
+                            // so we pick an appropriately sized integer type instead.
+                            //
+                            // NOTE: This is sub-optimal because in the case of (f32, f32, u32, u32)
+                            // we could do ([f32; 2], u64) which is better but this is the best we
+                            // can do right now.
+                            arg.cast_to(Reg { kind: RegKind::Integer, size });
+                        }
```

Comment:
Why not use the "homogeneous aggregate" concept that `rustc_target::abi::call` has?
(Unless you want to react to *any* floats being present in the aggregate, instead of it being *all* floats)

Reply:
If you're talking about [`rustc_target::abi::call::HomogeneousAggregate`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/abi/call/enum.HomogeneousAggregate.html) I already tried it in https://github.com/rust-lang/rust/pull/93564 and it gave some potentially non-negliable regressions albeit it was only in one crate, maybe things have improved things then. Do you want me to retry with it ?

Reply:
"Any" was not what I was targeting here, but it's a thing I also want to target at some point.